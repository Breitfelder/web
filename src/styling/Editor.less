.bblfsh-editor {
    &__container {
        height: 100%;
    }

    &__bookmark {
        display: inline-block;
        position: relative;
        width: 0;
        height: 0;

        &::before {
            content: '';
            position: absolute;
            bottom: -5px;
            left: -3px;
            width: 3px;
            height: 19px;
            border: 1px solid @border-accent;
            border-right-width: 0 !important;
            border-left-width: 2px !important;
        }
    }

    &__mark {
        background: @background-highlight;
    }
}
