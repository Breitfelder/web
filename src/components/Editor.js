import React, { Component } from 'react';
import styled, { injectGlobal } from 'styled-components';
import CodeMirror from 'react-codemirror';
import { background } from '../styling/variables';
import { font, border } from '../styling/variables';

import 'codemirror/lib/codemirror.css';
import 'codemirror/theme/solarized.css';
import 'codemirror/mode/python/python';
import 'codemirror/mode/clike/clike';

// eslint-disable-next-line
injectGlobal`
  .ReactCodeMirror, .CodeMirror {
    height: 100%;
  }

  .CodeMirror {
    font-family: ${font.family.code};
    font-size: ${font.size.large};
  }
`;

const Container = styled.div`
  height: 100%;
`;

const BookMark = styled.div`
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;

  &::before {
    content: '';
    position: absolute;
    bottom: -5px;
    left: -3px;
    width: 3px;
    height: 19px;
    border: 1px solid ${border.accent};
    border-right-width: 0 !important;
    border-left-width: 2px !important;
  }
`;

export default class Editor extends Component {
  get document() {
    return this.editor.getDoc();
  }

  get editor() {
    return this.refs.codemirror.getCodeMirror();
  }

  selectCode(from, to) {
    if (from && to) {
      return this.document.markText(from, to, {
        css: 'background: ' + background.highlight
      });
    } else if (from) {
      return this.document.setBookmark(from, { widget: this.bookmark });
    }
  }

  setMode(mode) {
    this.editor.setOption('mode', mode);
  }

  updateCode() {
    const cursor = this.cursor();
    this.document.setValue(this.props.code);
    this.document.setCursor(cursor);
  }

  onCursorActivity(editor) {
    if (!this.props.onCursorChanged) {
      return;
    }

    this.props.onCursorChanged(this.cursor(editor));
  }

  cursor(editor = this.editor) {
    return editor.getDoc().listSelections().slice(0, 1).pop().head;
  }

  render() {
    const { code, languageMode, onChange } = this.props;

    const options = {
      mode: languageMode,
      lineNumbers: true,
      theme: 'solarized light'
    };

    return (
      <Container>
        <CodeMirror
          ref="codemirror"
          autoFocus={true}
          onChange={onChange}
          value={code}
          onCursorActivity={editor => this.onCursorActivity(editor)}
          options={options}
        />
        <BookMark
          innerRef={elem => {
            this.bookmark = elem;
          }}
        />
      </Container>
    );
  }
}
