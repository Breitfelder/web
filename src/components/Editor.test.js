import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import Editor, { comparePos } from './Editor';

const mkPos = (line, ch) => ({ line, ch });

describe('Editor', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <Editor code='foo = 1' languageMode='python' />
    );

    expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
  });

  it('calls onCursorChanged when the selection changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onCursorChanged={spy} />,
      elem,
    );

    component.document.setSelection(
      mkPos(0, 4),      
      mkPos(0, 2),
    );

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const [ pos ] = spy.mock.calls[1];
    expect(pos).toEqual(mkPos(0, 2));
  });

  it('calls onCursorChanged when the cursor position changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onCursorChanged={spy} />,
      elem,
    );

    component.document.setCursor(mkPos(0, 4));

    // NOTE: there is an initial call when the CodeMirror is rendered.
    expect(spy.mock.calls.length).toBe(2);
    const [ pos ] = spy.mock.calls[1];
    expect(pos).toEqual(mkPos(0, 4));
  });

  it('calls onChange when the content changes', () => {
    const spy = jest.fn();
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' onChange={spy} />,
      elem,
    );

    component.document.replaceRange('changed', mkPos(0, 0), mkPos(0, 3));

    expect(spy.mock.calls.length).toBe(1);
    expect(spy.mock.calls[0][0]).toBe('changed = 1');
  });

  it('selects the code when selectCode is called', () => {
    const elem = document.createElement('div');
    const component = ReactDOM.render(
      <Editor code='foo = 1' languageMode='python' />,
      elem,
    );

    const assertMarks = n => expect(component.document.getAllMarks().length).toBe(n);

    assertMarks(0);

    const mark = component.selectCode(mkPos(0, 0), mkPos(0, 4));
    assertMarks(1);

    mark.clear();
    assertMarks(0);
  });
});
