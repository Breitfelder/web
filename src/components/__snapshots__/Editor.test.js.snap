// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders without crashing 1`] = `
<div
  className="bblfsh-editor__container"
>
  <Controlled
    autoFocus={true}
    onBeforeChange={[Function]}
    onCursor={[Function]}
    options={
      Object {
        "lineNumbers": true,
        "mode": "python",
        "theme": "solarized light",
      }
    }
    value="foo = 1"
  />
  <div
    className="bblfsh-editor__bookmark"
  />
</div>
`;
